### ISSUE TITLE
Enable and Configure Zone Aware Routing via xDS APIs rather than as a flag

### ISSUE THREAD

**Issue by liamawhite** [envoyproxy/envoy#2541]
**Title**:
Configure Zone Aware Routing via xDS APIs rather than as a flag

**Description**:
Currently to enable zone aware routing Envoy requires using the --service-zone flag. We (Istio) would like to be able to enable and configure zone aware routing via the xDS APIs. I believe this is achievable if the envoy instance looked itself up using the CDS API, retrieving a cluster array with just itself in.

**Comment by mattklein123** [comment]
Somewhat related to https://github.com/envoyproxy/envoy/issues/1344.

**Comment by cmluciano** [comment]
I'll take this on so that I can get more familiar with API changes

**Comment by htuch** [comment]
We might want to sync on this at some point soon. I'm going to be enabling locality (zonal) load balancing, which needs to have well defined behavior in the presence of zone-aware routing. The locality load balancing comes from existing EDS definitions.

**Comment by cmluciano** [comment]
@mattklein123 Talked with @liamawhite on this. The enhancement of allowing the zone to be supplied via the API call is so that we don't need to restart the container in Istio. 

@liamawhite Anything else I'm missing?

**Comment by liamawhite** [comment]
Specifically, its the restarting of Envoy. We have a preference for removing as much restarting of Envoy as possible.

**Comment by mattklein123** [comment]
@cmluciano @liamawhite I'm sorry I don't understand why Envoy needs to be restarted. Why would the zone of a container ever change once it is started?

**Comment by liamawhite** [comment]
Its a timing issue with Kube. When we schedule a pod we don't know which node it's going to run on. We can only retrieve the az of the node once the pod is running, therefore have to restart.

**Comment by mattklein123** [comment]
I don't know much about k8s, but it seems odd to me that you can't adjust the params of Envoy when the pod starts running? Potentially via an env var?

**Comment by liamawhite** [comment]
So you're suggesting not starting envoy until we can work out the zone?

**Comment by mattklein123** [comment]
@liamawhite yes that is what I'm suggesting. Basically all locality information should be available before starting Envoy IMO.

**Comment by cmluciano** [comment]
@liamawhite Think we can close this issue for now?

**Comment by mattklein123** [comment]
@cmluciano yeah I'm going to close as won't fix since I really don't think this is the right design direction. Let's reopen or discuss on envoy-dev@ if we think otherwise.

**Comment by liamawhite** [comment]
Agreed
